src/store/fs/util/entity_manager.rs

```diff
@@ -0,0 +1,1322 @@
+#![allow(dead_code)]
+use std::{fmt::Debug, future::Future, hash::Hash};
+
+use n0_future::{future, FuturesUnordered};
+use tokio::sync::{mpsc, oneshot};
+
+/// Trait to reset an entity state in place.
+///
+/// In many cases this is just assigning the default value, but e.g. for an
+/// `Arc<Mutex<T>>` resetting to the default value means an allocation, whereas
+/// reset can be done without.
+pub trait Reset: Default {
+    /// Reset the state to its default value.
+    fn reset(&mut self);
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum ShutdownCause {
+    /// The entity is shutting down gracefully because the entity is idle.
+    Idle,
+    /// The entity is shutting down because the entity manager is shutting down.
+    Soft,
+    /// The entity is shutting down because the sender was dropped.
+    Drop,
+}
+
+/// Parameters for the entity manager system.
+pub trait Params: Send + Sync + 'static {
+    /// Entity id type.
+    ///
+    /// This does not require Copy to allow for more complex types, such as `String`,
+    /// but you have to make sure that ids are small and cheap to clone, since they are
+    /// used as keys in maps.
+    type EntityId: Debug + Hash + Eq + Clone + Send + Sync + 'static;
+    /// Global state type.
+    ///
+    /// This is passed into all entity actors. It also needs to be cheap handle.
+    /// If you don't need it, just set it to `()`.
+    type GlobalState: Debug + Clone + Send + Sync + 'static;
+    /// Entity state type.
+    ///
+    /// This is the actual distinct per-entity state. This needs to implement
+    /// `Default` and a matching `Reset`. It also needs to implement `Clone`
+    /// since we unfortunately need to pass an owned copy of the state to the
+    /// callback - otherwise we run into some rust lifetime limitations
+    /// <https://github.com/rust-lang/rust/issues/100013>.
+    ///
+    /// Frequently this is an `Arc<Mutex<T>>` or similar. Note that per entity
+    /// access is concurrent but not parallel, so you can use a more efficient
+    /// synchronization primitive like [`AtomicRefCell`](https://crates.io/crates/atomic_refcell) if you want to.
+    type EntityState: Default + Debug + Reset + Clone + Send + Sync + 'static;
+    /// Function being called when an entity actor is shutting down.
+    fn on_shutdown(
+        state: entity_actor::State<Self>,
+        cause: ShutdownCause,
+    ) -> impl Future<Output = ()> + Send + 'static
```

Comment:
It's probably worth noting that all these `+ Send` bounds might make it a little harder to compile to Wasm.

Just something to keep an eye on.
Also FWIW, you're using `n0_future::future::Boxed`, but which will always assume `Send`, but there's also `n0_future::boxed::BoxFuture`, which is `Send` everywhere except Wasm, where it's `!Send`.

Reply:
Let's fix WASM once we get there. Or is there a SendUnlessWasm or something that I could just use here?